Common/Data/Custom/Fred/FredApi.cs

```diff
@@ -0,0 +1,149 @@
+        public int Count { get; set; }
+
+        [JsonProperty("offset")]
+        public int Offset { get; set; }
+
+        [JsonProperty("limit")]
+        public int Limit { get; set; }
+
+        [JsonProperty("observations")]
+        public IList<Observation> Observations { get; set; }
+
+        /// <summary>
+        /// Gets the FRED API token.
+        /// </summary>
+        public static string AuthCode { get; private set; } = string.Empty;
+
+        /// <summary>
+        /// Returns true if the FRED API token has been set.
+        /// </summary>
+        public static bool IsAuthCodeSet { get; private set; }
+
+        /// <summary>
+        /// Sets the EIA API token.
+        /// </summary>
+        /// <param name="authCode">The EIA API token</param>
+        public static void SetAuthCode(string authCode)
+        {
+            if (string.IsNullOrWhiteSpace(authCode)) return;
+
+            AuthCode = authCode;
+            IsAuthCodeSet = true;
+        }
+
+        /// <summary>
+        /// Return the URL string source of the file. This will be converted to a stream
+        /// </summary>
+        /// <param name="config">Configuration object</param>
+        /// <param name="date">Date of this source file</param>
+        /// <param name="isLiveMode">true if we're in live mode, false for backtesting mode</param>
+        /// <returns>
+        /// String URL of source file.
+        /// </returns>
+        public override SubscriptionDataSource GetSource(SubscriptionDataConfig config, DateTime date, bool isLiveMode)
+        {
+            return new SubscriptionDataSource(
+                $"https://api.stlouisfed.org/fred/series/observations?file_type=json&observation_start=1998-01-01&api_key={AuthCode}&series_id={config.Symbol}",
+                SubscriptionTransportMedium.Rest,
+                FileFormat.Collection);
+        }
+
+        /// <summary>
+        /// Readers the specified configuration.
+        /// </summary>
+        /// <param name="config">The configuration.</param>
+        /// <param name="content">The content.</param>
+        /// <param name="date">The date.</param>
+        /// <param name="isLiveMode">if set to <c>true</c> [is live mode].</param>
+        /// <returns></returns>
+        public override BaseData Reader(SubscriptionDataConfig config, string content, DateTime date, bool isLiveMode)
+        {
+            var series = JsonConvert.DeserializeObject<FredApi>(content).Observations;
+            var objectList = new List<FredApi>();
+            foreach (var observation in series)
+            {
+                decimal value;
+                if (decimal.TryParse(observation.Value, NumberStyles.Any, CultureInfo.InvariantCulture, out value))
+                {
+                    objectList.Add(new FredApi
+                    {
+                        Symbol = config.Symbol,
+                        Time = observation.Date - TimeSpan.FromDays(1),
```

Comment:
Hmmm, maybe it's forward adjusted so that no lookahead bias is introduced (same as comment above regarding the `Time` property) 🤔

Reply:
As per discussion in Slack (posting here to explain my comment since it's a little ambiguous right now):

So, because `new DateTime(yyyy, MM, dd)` sets the time to midnight by default if no Time component is passed, consider the following:

* Data from FRED, let's say VIX, is reported as `2019-09-29`, but what it actually refers to is `2019-09-29 17:00:00` (but they leave the time component out because it's daily data)

* We parse the date as `2019-09-29` because we have no time component (`new DateTime(2019, 9, 29);`)

* The `BaseData` instance will now have `2019-09-29 00:00:00` when the data was actually reported at `2019-09-29 17:00:00`

* The data point will be emitted at `2019-09-29 00:00:00` in backtests, basically getting us the value from the end of the day before it was even generated.
What I meant in that comment was that maybe they forward-adjust their times (think `Math.Ceil` but for time) so that if we parse it as `2019-09-30`, it will be for `2019-09-30 00:00:00` although the data point might actually be at `2019-09-29 17:00:00`

Reply:
The LEAN solution to this is to set a period for the data point and leave the Time property as the start time of the bar; the EndTime is automatically set to Time+Period.

Reply:
Great insight @gsalaz98!